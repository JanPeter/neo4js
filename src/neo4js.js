import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
