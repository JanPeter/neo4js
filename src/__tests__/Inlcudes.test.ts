import neo4js, {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
  IncludesQuery,
} from "../index";

type PersonProps = {
  name: string;
  active?: boolean;
};

type TaskProps = {
  title: string;
  done?: boolean;
};

class PersonModel extends Model<PersonProps, PersonInstance> {}
const Person: PersonModel = new PersonModel("Person");

class TaskModel extends Model<TaskProps, TaskInstance> {}
const Task: TaskModel = new TaskModel("Task");

const TaskCreatorRelation = relation
  .from(() => Person)
  .to(() => Task)
  .via("created");

const TaskAssigneeRelation = relation
  .from(Person)
  .to(Task)
  .via("assigned");

@model(Person)
class PersonInstance extends ModelInstance<PersonProps> {
  @hasMany(Task, TaskCreatorRelation)
  tasks: HasManyActions<TaskProps, TaskInstance>;

  @hasMany(Task, TaskAssigneeRelation)
  assignedTasks: HasManyActions<TaskProps, TaskInstance>;
}

@model(Task)
class TaskInstance extends ModelInstance<TaskProps> {
  @hasOne(() => Person, () => TaskCreatorRelation)
  creator: HasOneActions<PersonProps, PersonInstance>;
}

describe("Includes", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(async () => {
    const olaf = await Person.create({ name: "Olaf", active: true });
    let tasks = await olaf.assignedTasks.create(
      [
        { title: "Learn magic", done: true },
        { title: "Write more test cases 1", done: false },
        { title: "Write more test cases 2", done: true },
      ],
      { favorite: true }
    );
    for (const task of tasks) {
      await task.creator.add(olaf);
    }

    const ignatz = await Person.create({ name: "Ignatz", active: true });
    await ignatz.assignedTasks.create([
      { title: "Learn magic 2", done: true },
      { title: "Write more test cases 3", done: false },
      { title: "Write more test cases 4", done: true },
    ]);
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  describe("find with include", () => {
    it("should return olaf with all assigned tasks and the creators", async () => {
      let query = Person.findOneAndInclude({ name: "Olaf" });
      expect(query).toBeInstanceOf(IncludesQuery);

      if (query instanceof IncludesQuery) {
        let result = await query
          .include(o => o.assignedTasks.get({ done: true }))
          .include(r => r.creator.get())
          .run();

        expect(result).toBeInstanceOf(PersonInstance);
        expect(result.assignedTasks.length).toEqual(2);
        expect(result.assignedTasks[0]).toBeInstanceOf(TaskInstance);

        delete result.props.guid;
        result.assignedTasks = result.assignedTasks.map(t => {
          delete t.props.guid;
          if (t.creator) {
            delete t.creator.props.guid;
          }
          return t;
        });

        expect(result).toMatchSnapshot();
      } else {
        expect(true).toBeFalsy();
      }
    });

    it("should return all active users with done assigned tasks and the creators", async () => {
      let query = Person.findAndInclude({ active: true });
      expect(query).toBeInstanceOf(IncludesQuery);

      if (query instanceof IncludesQuery) {
        let result = await query
          .include(o => o.assignedTasks.get({ done: true }))
          .include(r => r.creator.get())
          .run();

        expect(result.length).toEqual(2);
        expect(result[0]).toBeInstanceOf(PersonInstance);
        expect(result[0].assignedTasks.length).toEqual(2);
        expect(result[0].assignedTasks[0]).toBeInstanceOf(TaskInstance);

        result = result
          .map(r => {
            delete r.props.guid;
            r.assignedTasks = r.assignedTasks.map(t => {
              delete t.props.guid;
              if (t.creator) {
                delete t.creator.props.guid;
              }
              return t;
            });
            return r;
          })
          .sort((a, b) => a.props.name.localeCompare(b.props.name));

        expect(result).toMatchSnapshot();
      } else {
        expect(true).toBeFalsy();
      }
    });
  });
});
