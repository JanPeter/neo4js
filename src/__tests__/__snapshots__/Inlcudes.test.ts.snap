// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Includes find with include should return all active users with done assigned tasks and the creators 1`] = `
Array [
  PersonInstance {
    "assignedTasks": Array [
      TaskInstance {
        "creator": PersonInstance {
          "assignedTasks": Object {
            "add": [Function],
            "count": [Function],
            "create": [Function],
            "get": [Function],
            "remove": [Function],
            "update": [Function],
          },
          "props": Object {
            "active": true,
            "name": "Olaf",
          },
          "tasks": Object {
            "add": [Function],
            "count": [Function],
            "create": [Function],
            "get": [Function],
            "remove": [Function],
            "update": [Function],
          },
        },
        "props": Object {
          "done": true,
          "title": "Write more test cases 2",
        },
        "relationProps": Object {
          "favorite": true,
        },
      },
      TaskInstance {
        "creator": PersonInstance {
          "assignedTasks": Object {
            "add": [Function],
            "count": [Function],
            "create": [Function],
            "get": [Function],
            "remove": [Function],
            "update": [Function],
          },
          "props": Object {
            "active": true,
            "name": "Olaf",
          },
          "tasks": Object {
            "add": [Function],
            "count": [Function],
            "create": [Function],
            "get": [Function],
            "remove": [Function],
            "update": [Function],
          },
        },
        "props": Object {
          "done": true,
          "title": "Learn magic",
        },
        "relationProps": Object {
          "favorite": true,
        },
      },
    ],
    "props": Object {
      "active": true,
      "name": "Olaf",
    },
    "tasks": Object {
      "add": [Function],
      "count": [Function],
      "create": [Function],
      "get": [Function],
      "remove": [Function],
      "update": [Function],
    },
  },
  PersonInstance {
    "assignedTasks": Array [
      TaskInstance {
        "creator": null,
        "props": Object {
          "done": true,
          "title": "Write more test cases 4",
        },
      },
      TaskInstance {
        "creator": null,
        "props": Object {
          "done": true,
          "title": "Learn magic 2",
        },
      },
    ],
    "props": Object {
      "active": true,
      "name": "Ignatz",
    },
    "tasks": Object {
      "add": [Function],
      "count": [Function],
      "create": [Function],
      "get": [Function],
      "remove": [Function],
      "update": [Function],
    },
  },
]
`;

exports[`Includes find with include should return olaf with all assigned tasks and the creators 1`] = `
PersonInstance {
  "assignedTasks": Array [
    TaskInstance {
      "creator": PersonInstance {
        "assignedTasks": Object {
          "add": [Function],
          "count": [Function],
          "create": [Function],
          "get": [Function],
          "remove": [Function],
          "update": [Function],
        },
        "props": Object {
          "active": true,
          "name": "Olaf",
        },
        "tasks": Object {
          "add": [Function],
          "count": [Function],
          "create": [Function],
          "get": [Function],
          "remove": [Function],
          "update": [Function],
        },
      },
      "props": Object {
        "done": true,
        "title": "Write more test cases 2",
      },
      "relationProps": Object {
        "favorite": true,
      },
    },
    TaskInstance {
      "creator": PersonInstance {
        "assignedTasks": Object {
          "add": [Function],
          "count": [Function],
          "create": [Function],
          "get": [Function],
          "remove": [Function],
          "update": [Function],
        },
        "props": Object {
          "active": true,
          "name": "Olaf",
        },
        "tasks": Object {
          "add": [Function],
          "count": [Function],
          "create": [Function],
          "get": [Function],
          "remove": [Function],
          "update": [Function],
        },
      },
      "props": Object {
        "done": true,
        "title": "Learn magic",
      },
      "relationProps": Object {
        "favorite": true,
      },
    },
  ],
  "props": Object {
    "active": true,
    "name": "Olaf",
  },
  "tasks": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
}
`;
